Workspaces

limethyst opened the pull request:
This pull request adds workspaces. It is not quite ready for merge yet, but functionality-wise it works just fine.

There are infinite workspaces that are indexed by numbers starting from 0. They are created automatically when you move to them or move a window to them. To open a workspace, I added the `yattac set-workspace` command and to move the currently focused window to a specified workspace I added `yattac move-window-to-workspace`.

Windows not on the current workspace are completely hidden (they don't appear in Alt+Tab or the taskbar).

The behavior with multiple monitors is that each monitor has its own set of workspaces and each monitor's workspace can be switched separately.

Also, this implentation of workspaces does not have support for workspaces as tags (having a window on multiple workspaces / having multiple workspaces open at the same time), but as far as I know bspwm does not have this functionality either and in my opinion implementing this would just be unnecessarily complicated for a feature that few people actually use.

There are still a few issues that need to be fixed before merge:
- [ ] Minimized windows appear on all workspaces
- [ ] There is no way to specify a default workspace for a window
- [ ] Weird behavior when switching windows virtual desktops
- [ ] When a window shows itself the workspace should automatically be switched to the one where the window is on (for example, if you click on a Firefox link, the Firefox window appears and then disappears immediately again, instead the workspace should be switched to the one Firefox is on)
- [x] If a window closes itself while not on the current workspace, resize adjustments may not be carried over (I could not find any way to produce this hypothetical bug so it probably doesn't exist)
- [x] Messy code
- [x] Restore focus on workspace switch
- [x] Add example hotkeys
- [x] When Yatta is closed with open windows on other workspaces, they become permanently irretrievable and can only be closed via Task Manager
- [x] Some programs (Firefox, IntelliJ on startup) may appear on multiple workspaces at once or just switch workspaces themselves
- [x] Minimizing windows leads to very weird behavior, such as making a different window appear on multiple workspaces at once

Fixes #14

Commit:
Add workspaces

limethyst commented:
While it seemed to be working without issues at first while testing earlier today, when using it all day today suddenly it became really buggy with windows randomly appearing on multiple workspaces, phantom windows etc.

I guess I need to do some more testing.

Maybe I should refactor the code so that each window only being present on once workspace is checked every time workspaces are switched.

limethyst commented:
Sometimes, Firefox seems to just not want to hide. It just doesn't disappear when switching to another Workspace. So far, I've only seen this with Firefox.

Commit:
Fix for windows randomly switching workspaces

Commit:
Fix for firefox sometimes becoming unmanaged when switching workspace

limethyst commented:
I have no idea what is wrong with Firefox, sometimes its window just isn't hidden by ShowWindow(SV_HIDE), but I've just added a couple of integrity checks here and there and now the issues seem to be gone.

Commit:
Refactoring

ghost commented:
The current behavior for minimized windows is that they appear on all workspaces in the taskbar and when unminimized they go into the current workspace. I tried storing minimized windows separately to make them appear only on their workspace, but failed as I could not find a way to reliably separate Window Hide events that are triggered by workspace switching from the ones triggered by minimizing.

Commit:
Show all windows on stop

Commit:
Add example hotkeys

Commit:
Added methods to directly access workspace fields

LGUG2Z commented:
First and foremost, thank you so much for having dedicated your time to understanding and contributing to this little project born out of my own idiosyncratic frustrations, I am humbled by the effort that you have put into this PR.

I will get this branch running locally towards the end of this week and start leaving review comments and suggestions this coming weekend. I'm very excited to work with you on getting this feature completed and merged in the future!

ghost commented:
Thank you for the kind words! This tiling window manager is the first one for windows I have tried so far that just works properly, doesn't feel too janky and isn't overly complicated to use/configure. I would love to get your feedback on the code quality when you have time, as I had to change a bit of code to make this work.

LGUG2Z commented on yatta/src/desktop.rs:
```diff
@@ -37,26 +37,137 @@ pub struct Desktop {
+                self.calculate_layout();
+                self.apply_layout(None);
+            }
+            self.verify(&mut Vec::new());
```
Currently when switching to a new workspace that has windows assigned to it, none of the associated windows will take the foreground focus and I need to use the mouse to focus one of them before I can start navigating again using the keyboard shortcuts.

Adding this snippet after the final `self.verify` call fixes this for me, always defaulting to giving focus to the first window on the workspace being switched to (this seems like a sensible default to me, but again I've never really been a heavy user of workspaces)

```rust
if let Some(window) = self.get_current_windows().first() {
    window.set_foreground()
}
```

limethyst replied:
As I mostly use the mouse for managing windows, I didn't really notice this, I will add this fix when I'm on my computer again, if you haven't by then. Just FYI you can add commits to my pull request too by just pushing to the master branch of my fork (by default, maintainers of the original repository get commit access to the fork when a pull request is made).

limethyst replied:
Also, better behavior might be to give back focus to the previously focused window.

limethyst replied:
As it turns out, it isn't that easy, as hwnd IDs change every time a window is hidden.

limethyst replied:
Oh, now I tried to do it by comparing titles, but as it turns out the yattac cmd window that appears for a split second is the foreground window when switching away :(

limethyst replied:
I have a little bit of ab implementation problem: Autohotkey's ", Hide" doesn't seem to completely hide yattac's window, so when switching away from the current workspace yattac is always the last focused window.

In the latest commit, I modified yattac to not have a window at all, but this also means that commands like yattac help have no output at all.

limethyst replied:
Also, as the hwnd's ID changes when hidden, I have to store the last focused window's title as a string, which could lead to issues when two windows have the same title.

LGUG2Z replied:
I've just tried running the latest build with these changes and I'm getting some funny behaviour where, although the `#![windows_subsystem = "windows"]` directive is supposed to suppress windows, now whenever `yattac` commands are getting run when I refresh my AHK config, a blue Powershell window is popping up every time. 

Once the key bindings have been set properly, I'm also not able to change focus reliably anymore (not too sure how to explain this, but if there is window beneath my focused window, focusing to the next window down will register an event but not update the display state, and when trying to swap windows up/down/left/right, one window will move but the other will not). Reverting back to an earlier build seems to solve these issues.

In light of the extra complexity of trying to revert back to the previously focused window on the last workspace, I'm happy to go with the more naive approach of defaulting to the 0th window in the list for each workspace upon switch, and opening a separate issue to track state across workspaces which could cover this and any other state-related refinements.

limethyst replied:
This simpler approach indeed seems good enough to me for this pull request, I removed the windows-subsystem directive and implemented your behavior.

Commit:
Remove unnecessary verification

This was added as a lazy temporary fix for some bugs that isn't necessary anymore.

LGUG2Z commented on yatta/src/main.rs:
```diff
@@ -630,6 +625,20 @@ fn handle_socket_message(
+                        SocketMessage::SetWorkspace(index) => {
+                            d.set_workspace(index);
+                        }
+                        SocketMessage::MoveWindowToWorkspace(index) => {
```
I would like to have an accompanying `MoveWindowToWorkspaceAndFollow` message since this is what "feels" like the natural behaviour to me. The handler would be the same as here with the addition of this at the end:

```rust
d.set_workspace(index);
```

For the correct window to be focused when following a moved window to its new workspace, we would need to modify the `WindowsEventType::Show` handler in `main.rs` to explicitly set the window emitting the show event to be the new foreground window:

```rust
let idx = display.get_foreground_window_index() + 1;
// If we are inserting where there is a window that has resize adjustments, take
// over those resize adjustments and remove them from the window that is
// currently there
if let Some(current_window) = display.get_current_windows_mut().get_mut(idx) {
    let resize = current_window.resize.clone();
    current_window.resize = None;
    ev.window.resize = resize;
}

display.get_current_windows_mut().insert(idx, ev.window);
display.calculate_layout();
display.apply_layout(None);
ev.window.set_foreground(); // <-- explicitly set foreground of the newly shown window on the workspace here

if let Some(title) = ev.window.title() {
    if let Ok(path) = ev.window.exe_path() {
        info!(
            "managing new window: {} - {} ({})",
            &exe_name_from_path(&path),
            &title,
            ev.window.hwnd.0
        );
    }
}
```

limethyst replied:
This seems like a great idea! Not something that I would use, as I move windows simply to remove then when they are cluttering my current workspace, but I'll implement this right now.

Commit:
Restore focus on workspace switch

Commit:
Remove temporary print statement

LGUG2Z commented on yatta/src/main.rs:
```diff
@@ -321,10 +314,12 @@ fn handle_windows_event_message(mut ev: WindowsEvent, desktop: Arc<Mutex<Desktop
                         }
                     }
                 }
+            } else if !display.get_workspace().windows.contains(&ev.window) {
```
I have been playing around with this part trying to get the following situation working:

Workspace 1 has an application open with a clickable link
Workspace 2 has Firefox
When I click the link in the application running on Workspace 1
Then I want to be directed to the newly opened tab in Firefox in Workspace 2

This code gets me to the right workspace, but it won't yet reliably ensure that the browser is the focused window:

```rust
} else if !display.get_workspace().windows.contains(&ev.window) {
    let mut workspace_idx = None;

    for (i, w) in display.workspaces.iter().enumerate() {
        if w.windows.contains(&ev.window) {
            workspace_idx = Option::from(i);
        }
    }

    if let Some(w_idx) = workspace_idx {
        display.set_workspace(w_idx);
    }
}
```

limethyst replied:
Unfortunately, this solution reintruduces a couple of problems that I fixed by adding a window.hide in there: Some programs (like Firefox) when hidden sometimes just don't hide or reshow immediately, and this code makes these programs just sometimes immediately switch back to their workspace when switching away from them.

limethyst replied:
I mean, given that problem, there might be no way to implement this at all.

LGUG2Z replied:
I have modified that snippet a little more to invoke the hiding when a workspace index can't be determined:

```rust
match workspace_idx {
    None => ev.window.hide(),
    Some(idx) => display.set_workspace(idx),
}
```

As far as I can tell this hasn't introduced any issues with hiding and reshowing with Firefox across workspaces and allows for following the focus of links opened in a browser that is in a different workspace.

When you get a chance can you try running a build with this for a little while and reporting how it goes for you?

I'm also going to keep this build running for the rest of the day and I'll report if any related issues arise.

limethyst replied:
Hmm... While this doesn't seem like it would fix the issue, the fact that you haven't experienced this issue makes me think that it might have been caused by something else. I will try this version and tell if I have experienced any issues with it.

limethyst replied:
> 
> 
> I have been playing around with this part trying to get the following situation working:
> 
> Workspace 1 has an application open with a clickable link
> Workspace 2 has Firefox
> When I click the link in the application running on Workspace 1
> Then I want to be directed to the newly opened tab in Firefox in Workspace 2
> 
> This code gets me to the right workspace, but it won't yet reliably ensure that the browser is the focused window:
> 
> ```rust
> } else if !display.get_workspace().windows.contains(&ev.window) {
>     let mut workspace_idx = None;
> 
>     for (i, w) in display.workspaces.iter().enumerate() {
>         if w.windows.contains(&ev.window) {
>             workspace_idx = Option::from(i);
>         }
>     }
> 
>     if let Some(w_idx) = workspace_idx {
>         display.set_workspace(w_idx);
>     }
> }
> ```

Just FYI there is an `iter.position` command, which can be used to write things like these in one line.

limethyst replied:
I tried implementing your code, but as I sometimes instead of switching to another workspace other windows would now sometimes switch to the current workspace when switching from a workspace with firefox in it.

LGUG2Z replied:
Do you think you can come up with a minimal reproducible example of a start state of workspaces and a sequence of events that you can use to reliably produce the behaviour you're talking about? If you can do a little screen recording of it as well that would be amazing. 

I haven't seen any funky today, but it might be that I am just not registering the same thing that you are seeing as weird behaviour. 😅

limethyst replied:
That seems like a good idea and I will try to do so tomorrow, but it might not be easy as it doesn't happen often and seems to just be random.

limethyst replied:
Also, it only happens with Firefox.

LGUG2Z replied:
It happened to me this morning! Though I could not tell you for the life of me the exact sequence of events that produced the behaviour. I'm happy to leave this as-is for now and open a separate issue for trying to get focus changes to switch to other workspaces when applicable in the future.

Commit:
Add MoveWindowToWorkspaceAndFollow

ghost commented:
Some programs like IntelliJ like to steal focus when their windows are reshown. This means that when switching to a workspace with MoveWindowToWorkspaceAndFollow which has IntelliJ on it the window won't be focused. I don't know this happens yet.

Commit:
Remove windows_subsystem directive

Commit:
Set focus to first window on workspace change

LGUG2Z commented:
> 
> 
> Some programs like IntelliJ like to steal focus when their windows are reshown. This means that when switching to a workspace with MoveWindowToWorkspaceAndFollow which has IntelliJ on it the window won't be focused. I don't know this happens yet.

Recreated this behaviour this morning and the focus stealing of `idea64.exe` is indeed plain in the logs. I can live with this since I spend most of my time on the computer staring at IntelliJ anyway 😅

```
[2021-07-15 07:08:10.919961 -07:00] INFO [yatta] yatta\src\main.rs:439: handling yattac socket message: MoveWindowToWorkspaceAndFollow(0)
[2021-07-15 07:08:10.946721 -07:00] DEBUG [yatta::window] yatta\src\window.rs:309: managing firefox.exe - Workspaces by limethyst A· Pull Request #15 A· LGUG2Z/yatta â?" Mozilla Firefox (styles: BORDER | CAPTION | CLIPCHILDREN | CLIPSIBLINGS | DLGFRAME | GROUP | MAXIMIZEBOX | MINIMIZEBOX | OVERLAPPEDWINDOW | SIZEBOX | SYSMENU | TABSTOP | THICKFRAME | TILEDWINDOW | VISIBLE) (extended styles: WINDOWEDGE)
[2021-07-15 07:08:10.946991 -07:00] DEBUG [yatta::window] yatta\src\window.rs:309: managing idea64.exe - yatta â?" desktop.rs (styles: BORDER | CAPTION | CLIPCHILDREN | CLIPSIBLINGS | DLGFRAME | GROUP | MAXIMIZEBOX | MINIMIZEBOX | OVERLAPPEDWINDOW | SIZEBOX | SYSMENU | TABSTOP | THICKFRAME | TILEDWINDOW | VISIBLE) (extended styles: WINDOWEDGE)
[2021-07-15 07:08:10.948473 -07:00] DEBUG [yatta::window] yatta\src\window.rs:309: managing idea64.exe - yatta â?" desktop.rs (styles: BORDER | CAPTION | CLIPCHILDREN | CLIPSIBLINGS | DLGFRAME | GROUP | MAXIMIZEBOX | MINIMIZEBOX | OVERLAPPEDWINDOW | SIZEBOX | SYSMENU | TABSTOP | THICKFRAME | TILEDWINDOW | VISIBLE) (extended styles: WINDOWEDGE)
[2021-07-15 07:08:10.948915 -07:00] DEBUG [yatta::window] yatta\src\window.rs:309: managing WindowsTerminal.exe - Windows PowerShell (styles: BORDER | CAPTION | CLIPSIBLINGS | DLGFRAME | GROUP | MAXIMIZEBOX | MINIMIZEBOX | OVERLAPPEDWINDOW | SIZEBOX | SYSMENU | TABSTOP | THICKFRAME | TILEDWINDOW | VISIBLE) (extended styles: WINDOWEDGE)
[2021-07-15 07:08:10.961074 -07:00] INFO [yatta] yatta\src\main.rs:165: handling yatta channel message: FocusChange (SystemForeground)
[2021-07-15 07:08:10.979628 -07:00] DEBUG [yatta::window] yatta\src\window.rs:309: managing firefox.exe - Workspaces by limethyst A· Pull Request #15 A· LGUG2Z/yatta â?" Mozilla Firefox (styles: BORDER | CAPTION | CLIPCHILDREN | CLIPSIBLINGS | DLGFRAME | GROUP | MAXIMIZEBOX | MINIMIZEBOX | OVERLAPPEDWINDOW | SIZEBOX | SYSMENU | TABSTOP | THICKFRAME | TILEDWINDOW | VISIBLE) (extended styles: WINDOWEDGE)
[2021-07-15 07:08:10.983583 -07:00] INFO [yatta] yatta\src\main.rs:370: focusing window: firefox.exe - Workspaces by limethyst A· Pull Request #15 A· LGUG2Z/yatta â?" Mozilla Firefox (263250)
[2021-07-15 07:08:10.993119 -07:00] INFO [yatta] yatta\src\main.rs:165: handling yatta channel message: FocusChange (SystemForeground)
[2021-07-15 07:08:11.001206 -07:00] INFO [yatta] yatta\src\main.rs:370: focusing window: idea64.exe - yatta â?" desktop.rs (3540886)
[2021-07-15 07:08:11.009771 -07:00] DEBUG [yatta::window] yatta\src\window.rs:309: managing idea64.exe - yatta â?" desktop.rs (styles: BORDER | CAPTION | CLIPCHILDREN | CLIPSIBLINGS | DLGFRAME | GROUP | MAXIMIZEBOX | MINIMIZEBOX | OVERLAPPEDWINDOW | SIZEBOX | SYSMENU | TABSTOP | THICKFRAME | TILEDWINDOW | VISIBLE) (extended styles: WINDOWEDGE)
```

ghost commented:
Apparently, this is a problem for Linux tiling window managers too: https://youtrack.jetbrains.com/issue/IDEA-204227

LGUG2Z committed:
Merge branch 'master' into limethyst/master

LGUG2Z committed:
style(clippy): run +nightly fmt --all and apply clippy lints

LGUG2Z commented:
I just pushed the latest changes from the `master` branch onto your PR (with the fix for Steam) and then another commit where I ran `cargo +nightly fmt --all` and applied all but one of the suggestions from `cargo clippy`. This is the clippy suggestion that I am unsure of, it feels like it would be making a logic change, but I can't quite put my finger on how/why. I'd appreciate if you could have a look over it and see if you can grok what I'm missing:

```
❯ cargo clippy
    Checking yatta v0.1.0 (C:\Users\Jade Iqbal\IdeaProjects\yatta\yatta)
warning: all if blocks contain the same code at the end
   --> yatta\src\main.rs:306:17
    |
306 | /                     display.calculate_layout();
307 | |                     display.apply_layout(None);
308 | |
309 | |                     if let Some(title) = ev.window.title() {
...   |
318 | |                     }
319 | |                 }
    | |_________________^
    |
    = note: `#[warn(clippy::branches_sharing_code)]` on by default
    = warning: Some moved values might need to be renamed to avoid wrong references
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#branches_sharing_code
help: consider moving the end statements out like this
    |
306 |                 }
307 |                 display.calculate_layout();
308 |                 display.apply_layout(None);
309 |
310 |                 if let Some(title) = ev.window.title() {
311 |                     if let Ok(path) = ev.window.exe_path() {
  ...

warning: 1 warning emitted

    Finished dev [unoptimized + debuginfo] target(s) in 1.00s
```

ghost commented:
Hmmm... For some reason this lint does not show up when I run cargo clippy (or cargo +nightly clippy)

ghost commented:
Also, this lint is known to cause logic changes sometimes https://github.com/rust-lang/rust-clippy/issues/7452

LGUG2Z commented on yatta/src/main.rs:
```diff
@@ -631,6 +633,24 @@ fn handle_socket_message(
+                        }
+                        SocketMessage::Stop => {
+                            let windows = desktop.get_all_windows();
+                            for mut window in windows {
```
Can you also work this unhiding behaviour into a ctrl+c signal hook? When I'm developing I tend to run `cargo run` in the terminal and watch the logs and I have a tendency to ctrl+c out to restart the proc instead of running `yattac stop` from a different terminal, and on more than one occasion I've been left with invisible windows I had to kill with `Stop-Process` and then restart to regain visibility.

LGUG2Z replied:
Addressed this in my latest commit to your fork with a a ctrl-c handler that listens on a channel in a dedicated thread.

LGUG2Z committed:
style(clippy): allow clippy::branches_sharing_code on false positive

LGUG2Z committed:
feat(workspaces): restore all hidden windows on ctrl-c event

LGUG2Z commented:
In terms of functionality and implementation I am happy with the state of the PR as it stands now. Let me know how you feel about it, if you still want to change/add anything, add comments etc., and hopefully we can resolve anything outstanding and merge this to the master branch soon! @limethyst

ghost commented:
I would be happy with this pull request being merged as is. While there are a few workspace-related features I would still like to add in the future the workspace functionality in this pull request is sufficient for me. I also haven't found any unresolved bugs yet.

I'm currently on a vacation, which is why I haven't added any commits over the past week, but when I'm back home I will probably submit more pull requests adding some of these features.

ghost commented:
I did find a bug.
Electron apps' rendering seems to hang when switching away from and to their workspace. This is fixed by simply resizing or moving the window.

ghost commented:
This also happens with websites opened in Google Chrome.

ghost commented:
I made a screen recording of the bug:

https://user-images.githubusercontent.com/23194836/126308528-1dde6042-2501-49e0-8728-342b447254ca.mp4

I need to find a way to send a "window update" event without actually resizing or moving the window.

ghost commented:
This is suprisingly tricky to fix. Directly sending a resize message does not work, resizing the window by one pixel and then back immediately only works sometimes. I guess a delay is needed to give chrome time to handle the event, but that would look quite janky.

I can force the window to repaint once by sending it a Repaint message, but that only updates once and the window still does not update on any subsequent interactions.

LGUG2Z commented:
I have been able to reproduce this bug a few times with Google Chrome, but not reliably, and (so far) only ever when the window is the only window on a workspace.

Since this is only impacting Chromium applications, is there a chance that this could be related to an upstream Chromium issue on Windows?

ghost commented:
I can reliably reproduce this issue, here's how:
- Open chrome on Workspace 2
- Open another window (e.g. IntelliJ) on workspace 1
- Switch to Workspace 2 and use interact with the Chrome window
- Switch to Workspace 1 and interact with the window there (like IntelliJ)
- Switch to Workspace 2, click on Chrome and interact with the currently open tab

For me, this also causes the bug to occur when Chrome is not the only window on the workspace.

Also, this might be an upstream "issue" with Chrome, but that depends on whether they consider weird behavior triggered by weird Windows API use to be a bug. In electron, you are for example not supposed to hide the window directly but rather call BrowserWindow.hide() or .show() from the JavaScript code inside, which handles this. (There are bugs that trigger the same behavior from inside Electron though, see https://github.com/electron/electron/issues/27353 and https://github.com/electron/electron/issues/28255)

ghost commented:
Also, this behavior might not happen if background throttling (a chrome & electron feature) is disabled, I haven't tested this yet though.

ghost commented:
I might submit a bug report to Chromium, however a workaround would still be needed as
1. Chrome has tons of open issues and I imagine one like this might have a very low probability
2. All electron programs will still have to upgrade to the new electron version
3. This bug is very annoying for someone who needs to use a lot of electron apps (like me)

limethyst commented:
The nuclear option would be to inject a DLL into Electron processes and Chrome to call the internal hide function.

LGUG2Z commented:
https://github.com/workspacer/workspacer/issues/171 Do you think that this is the same issue? From reading some of the comments in that thread it feels similar to what we are experience on workspace changes.

ghost commented:
While this issue is a different one, one of the devs mentions pretty much this exact issue in a comment on this issue https://github.com/workspacer/workspacer/issues/171#issuecomment-759342377

ghost commented:
One of the devs proposed to replace the workspace system with windows virtual desktops there because of the various issues with hiding and showing programs 😕 https://github.com/workspacer/workspacer/issues/219

ghost commented:
While using windows virtual desktops caused numerous issues before, I could maybe implement a "mapping" of sorts of yatta workspaces and Windows virtual desktops.

ghost commented:
I found [this](https://microsoft.github.io/windows-docs-rs/doc/bindings/Windows/Win32/UI/Shell/struct.IVirtualDesktopManager.html) fairly rudimentary API for managing virtual desktops

LGUG2Z commented:
I actually had an implementation of workspaces using virtual desktops that worked well for everything except moving windows between virtual desktops reliably: https://github.com/LGUG2Z/yatta/commit/65414f302b5c7ed266a1eed380b5e002d443ce0e

If it's possible to figure out the window movement between virtual desktops without making the redrawing go crazy, we could also re-use the old code that was used to configure assignments of specific applications to specific virtual desktops.

ghost commented:
Does [MoveWindowToDesktop](https://microsoft.github.io/windows-docs-rs/doc/bindings/Windows/Win32/UI/Shell/struct.IVirtualDesktopManager.html#method.MoveWindowToDesktop) make redrawing go crazy?

LGUG2Z commented:
I think the issue was probably more in the spaghetti logic of ensuring desktop rules that was specific to `yatta` now that I think about it. At the time, `windows-rs` hadn't yet exposed that API so I used https://crates.io/crates/winvd for all of the virtual desktop interactions. Given that the API is now exposed in an official Microsoft crate I think it's definitely worth giving this approach a second look!

ghost commented:
The way I would go about implementing this would be like this:
- Store windows and layout for each virtual desktop like it is stored right now in this pull request (to avoid recalculating layout every time on workspace switch)
- Store the workspace number and if it changed after any WindowsSocketMessage handle the workspace change by checking if all windows are still open and recalculating layout if needed

ghost commented:
Unfortunately, there is no Workspace Switch event, but that could just be handled by checking the current workspace on a window show or hide event.

ghost commented:
There is one really big problem I have with windows virtual desktops: They are REALLY annoying with multiple monitors, as all monitors are always on the same workspace (you can't have one fixed window open on one monitor and switch workspaces on the other).

LGUG2Z commented:
https://github.com/electron/electron/pull/29919 this PR for a bug in Electron might be relevant (also references potential regression that introduced this behaviour in Chromium) - I have noticed when reproducing this that the content is being loaded but not painted.

My own little workaround for this at the moment is to just hit ALT+SHIFT+R to call `Run, yattac.exe retile, Hide` in my `yatta.ahk` whenever I see it happen. I also tried changing my workspace-switching bindings to this, which significantly reduced the occurrences of this behaviour (though, did not eliminate it completely):

```ahk
!1::
Run, yattac.exe set-workspace 0, Hide
Run, yattac.exe retile, Hide
return
```

The downside to the latter approach is of course that doubling up on the retiling on workspace switching seems "janky".

LGUG2Z committed:
Merge branch 'master' into limethyst/master

LGUG2Z commented:
@limethyst You may be interested to know that after our discussions I set out to write a new tiling window manager for WIndows 10 based on the learnings from `yatta`, you can take a look at it here if you are interested: https://github.com/LGUG2Z/komorebi/ (no readme yet)

Very similar architecture at a high level, but a much cleaner implementation that is built from the ground-up to support multiple monitors, workspaces and stacks.

I will be on holiday for a while but once I return I will start documenting it and opening issues asking for help on things that I am technically unable to come up with an implementation for myself.

Thank you for the huge burst of inspiration!

ghost commented:
Starting a new project sounds like a great idea! I've just come back from vacation (great timing lol) and I might contribute to the new project too. So do you want to just abandon this pull request and stop updating this project in favor of the new one or would you still try to get this one working? (To be honest, I'd rather abandon this one if it's going to be replaced anyway) @LGUG2Z

ghost commented:
Also, the name komorebi is already used by this project (well-known at 2300 stars, I've even used it myself): https://github.com/cheesecakeufo/komorebi

The other project does something different and is for Linux only, but could still be confusing when googling.

LGUG2Z commented:
I hope you enjoyed your vacation!

Let's go ahead and abandon this PR. I'll probably archive this repo in the future as an example of an MVP of a tiling window manager that was built to handle a single monitor and a single screen as a learning resource (and how the choice of data structures made more difficult the implementation of multi-monitor, multi-workspace and container support).

Although `komorebi` doesn't have a readme yet, the config is basically the same as `yatta`; an AHK script with mostly the same (and a bunch of new commands) in `komorebic`. The workspaces implementation is roughly the same as the one in this PR in that it doesn't try to integrate the Windows Virtual Desktops feature, so it will likely still have the same issue with Chromium apps.

As for the name, it was my wife's choice (and her first time naming one of my projects!) so I'll keep it as-is despite the Linux project of the same name, Like you say, it's for a different platform and a different use case, and if other tiling window manager nerds are anything like me, they'll usually append "tiling window manager" to the end of their Google searches 99% of the time. 😅

I'd be very happy to work with you again on this or any other projects of yours that you think I might find interesting!
